### PR TITLE
Communication Email job to create Communication Email records for Communication

### PR DESCRIPTION
**Pull request by leobcgov**
## Status
**Ready**

## Fixes or Implements [AB#8638](https://qslvic.visualstudio.com/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/8638)
- [X] Enhancement
- [ ] Bug
- [ ] Documentation

## Description:
- Communication Email job to create Communication Email records for Communication records which are not in Processed status
- This PR includes the improvement of a single LINQ query to retrieve User Profiles.
- This PR doesn't include a translation of LINQ to LAMBDA due to the challenge of conversion. Per MS's official doc & guidance, I suggest using LINQ is ok for our project for the complex queries <https://docs.microsoft.com/en-us/ef/core/querying/complex-query-operators> and <https://github.com/dotnet/EntityFramework.Docs>. This can be discussed on our daily call.

## Testing
- [ ] Unit Tests Updated
- [ ] Functional Tests Updated
- [X] Not Required

### Steps to Reproduce:
If this is a bug, please provide details on how to reproduce the code.

### UI Changes
NO

### Browsers Tested
- [ ] Chrome
- [ ] Safari
- [ ] Edge
- [ ] Firefox

Provide an explanation for any test(s) not completed.

## Notes/Additional Comments
Any addtional notes or comments that may be relevant.  Include any specialized deployment steps here.

### CODE REVIEW

File: Apps/Database/src/Models/UserProfile.cs [hunk 1]
```diff
@@ -72,5 +73,10 @@ public class UserProfile : AuditableEntity
         /// </summary>
         [MaxLength(44)]
         public string? EncryptionKey { get; set; }
+
```
Comment (sslaws) [inline Apps/Database/src/Models/UserProfile.cs:76]: We don’t need this at this point.

File: Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            this.logger.LogDebug($"Finished adding Communication in DB");
+            return result;
+        }
+
+        /// <inheritdoc />
+        public List<UserProfile> GetActiveUserProfilesByCommunicationId(Guid communicationId, int maxRows)
+        {
+            this.logger.LogTrace($"Getting active user profiles to send emails from DB...");
+
+            var userProfiles = (from profile in this.dbContext.UserProfile
```
Comment (sslaws) [inline Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs:84]: I don’t like this as it’s the only location where we use this syntax. Did the query I supply as a base not work?

File: Apps/Database/src/Models/Email.cs [hunk 3]
```diff
@@ -99,5 +100,10 @@ public class Email : AuditableEntity
         [Required]
         [MaxLength(10)]
         public string EmailStatusCode { get; set; } = EmailStatus.New;
+
+        /// <summary>
+        /// Gets the list containing all the communication emails related to this email.
+        /// </summary>
+        public virtual List<CommunicationEmail> CommunicationEmails { get; } = new List<CommunicationEmail>();
```
Comment (sslaws) [inline Apps/Database/src/Models/Email.cs:107]: We don’t need this here

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                foreach (Communication communication in communications)
+                {
+#pragma warning disable CA1031 //We want to catch exception.
+                    try
+                    {
+                        List<UserProfile> usersToSendCommEmails = new List<UserProfile>();
+                        bool moreUsersToCreateCommunicationEmails = false;
+                        do
+                        {
+                            usersToSendCommEmails = this.commEmailDelegate.GetActiveUserProfilesByCommunicationId(communication.Id, this.maxFetchSize);
```
Comment (sslaws) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:89]: The method should be more descriptive like getProfilesForCommunication

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+                            if (communication.CommunicationStatusCode != CommunicationStatus.Processing)
+                            {
+                                communication.CommunicationStatusCode = CommunicationStatus.Processing;
+                                this.communicationDelegate.Update(communication, false);
+                            }
+
+                            this.dbContext.SaveChanges(); // commit after every retryFetchSize (or 250) pairs of Email & CommunicationEmail.
+
+                            moreUsersToCreateCommunicationEmails = usersToSendCommEmails.Count == this.maxFetchSize ? true : false;
```
Comment (sslaws) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:122]: The ? Operator is unnecessary

File: Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs (hunk 2 above)
Comment (leobcgov) [reply Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs:84]: I have a hard time to make it work. I can look further into the conversion of the LINQ to Lambda if needed.

File: Apps/Database/src/Models/UserProfile.cs (hunk 1 above)
Comment (leobcgov) [reply Apps/Database/src/Models/UserProfile.cs:76]: Do you mean the empty line or the EncryptionKey  ?

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 5 above)
Comment (leobcgov) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:122]: you are right. I 'm deleting " ? true : false;".

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 4 above)
Comment (leobcgov) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:89]: renaming..

File: Apps/Database/src/Models/Email.cs (hunk 3 above)
Comment (leobcgov) [reply Apps/Database/src/Models/Email.cs:107]: removed.

File: Apps/Database/src/Context/GatewayDbContext.cs [hunk 6]
```diff
@@ -687,6 +688,15 @@ private void SeedCommunication(ModelBuilder modelBuilder)
                     UpdatedDateTime = this.DefaultSeedDate,
                 },
                 new CommunicationStatusCode
+                {
+                    StatusCode = CommunicationStatus.Processing,
+                    Description = "Communication is being processed by batch",
```
Comment (FuriousLlama) [inline Apps/Database/src/Context/GatewayDbContext.cs:693]: nit: Communication is being processed

File: Apps/Database/src/Models/UserProfile.cs [hunk 7]
```diff
@@ -72,5 +73,10 @@ public class UserProfile : AuditableEntity
...
+        /// <summary>
+        /// Gets the list containing all the communication emails related to this user profile.
+        /// </summary>
+        public virtual List<CommunicationEmail> CommunicationEmails { get; } = new List<CommunicationEmail>();
```
Comment (FuriousLlama) [inline Apps/Database/src/Models/UserProfile.cs:80]: I don't think we have the other objects on the profile table. Is this necessary to get the foreign key relationship to work?

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        /// </summary>
+        /// <param name="configuration">The configuration to use.</param>
+        /// <param name="logger">The logger to use.</param>
+        /// <param name="communicationDelegate">The Communication delegate to use.</param>
+        /// <param name="commEmailDelegate">The Communication Email delegate to use.</param>
+        /// <param name="emailQueueService">The email queue service to use.</param>
+        /// <param name="dbContext">The db context to use.</param>
+        public CommunicationJob(IConfiguration configuration, ILogger<CommunicationJob> logger, ICommunicationDelegate communicationDelegate, ICommunicationEmailDelegate commEmailDelegate, IEmailQueueService emailQueueService, GatewayDbContext dbContext)
+        {
+            Contract.Requires((configuration != null) && (emailQueueService != null));
```
Comment (FuriousLlama) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:56]: de don't use contracts anymore. Making the objects non-nullable removes the need for them.

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+                                CommunicationEmail commEmail = new CommunicationEmail()
+                                {
+                                    Email = email,
+                                    UserProfile = profile,
+                                    Communication = communication,
+                                };
+                                this.commEmailDelegate.Add(commEmail, false);
+                            }
+
+                            if (communication.CommunicationStatusCode != CommunicationStatus.Processing)
```
Comment (FuriousLlama) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:114]: this only needs to be done once per the parent loop. Let's move it to line 85 or before the foreach UserProfile

Comment (FuriousLlama) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:114]: For readability, how about moving this if to before the foreach on the profile list. Say line 89

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+                            {
+                                communication.CommunicationStatusCode = CommunicationStatus.Processing;
+                                this.communicationDelegate.Update(communication, false);
+                            }
+
+                            this.dbContext.SaveChanges(); // commit after every retryFetchSize (or 250) pairs of Email & CommunicationEmail.
```
Comment (FuriousLlama) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:120]: nit: no need for the comment, or move above the line

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+                            moreUsersToCreateCommunicationEmails = usersToSendCommEmails.Count == this.maxFetchSize;
+                        }
+                        while (moreUsersToCreateCommunicationEmails); // keep looping when the above query returns the max return rows (or user profiles).
```
Comment (FuriousLlama) [inline Apps/JobScheduler/src/Jobs/CommunicationJob.cs:124]: nit: no need for the comment

**Comment by FuriousLlama** [comment]
Looks much, much cleaner Leo!

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 11 above)
Comment (leobcgov) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:124]: updated

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 9 above)
Comment (leobcgov) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:114]: great point!

File: Apps/JobScheduler/src/Jobs/CommunicationJob.cs (hunk 8 above)
Comment (leobcgov) [reply Apps/JobScheduler/src/Jobs/CommunicationJob.cs:56]: removed the Contract line

File: Apps/Database/src/Models/UserProfile.cs (hunk 7 above)
Comment (leobcgov) [reply Apps/Database/src/Models/UserProfile.cs:80]: removing it to try now.

File: Apps/Database/src/Context/GatewayDbContext.cs (hunk 6 above)
Comment (leobcgov) [reply Apps/Database/src/Context/GatewayDbContext.cs:693]: updated

File: Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs (hunk 2 above)
Comment (leobcgov) [reply Apps/Database/src/Delegates/DBCommunicationEmailDelegate.cs:84]: @TiagoGraf89  has found a winning lambda to replace this Linq.

File: Apps/Database/src/Models/UserProfile.cs (hunk 1 above)
Comment (leobcgov) [reply Apps/Database/src/Models/UserProfile.cs:76]: removing virtual ... CommunicationEmails

**Comment by leobcgov** [comment]
@ all: please can you re-review this PR?

File: Apps/Database/src/Models/UserProfile.cs [hunk 12]
```diff
@@ -16,6 +16,7 @@
 namespace HealthGateway.Database.Models
 {
     using System;
+    using System.Collections.Generic;
```
Comment (FuriousLlama) [inline Apps/Database/src/Models/UserProfile.cs:19]: is this necessary?

File: Apps/Database/src/Models/Email.cs [hunk 13]
```diff
@@ -16,6 +16,7 @@
 namespace HealthGateway.Database.Models
 {
     using System;
+    using System.Collections.Generic;
```
Comment (FuriousLlama) [inline Apps/Database/src/Models/Email.cs:19]: no longer necessary

File: Apps/Database/src/Models/Communication.cs [hunk 14]
```diff
@@ -16,6 +16,7 @@
 namespace HealthGateway.Database.Models
 {
     using System;
+    using System.Collections.Generic;
```
Comment (FuriousLlama) [inline Apps/Database/src/Models/Communication.cs:19]: no longer necessary

Comment (leobcgov) [reply Apps/Database/src/Models/Communication.cs]: removed from my last checkin

File: Apps/Database/src/Models/Email.cs (hunk 13 above)
Comment (leobcgov) [reply Apps/Database/src/Models/Email.cs:19]: removed from my last check-in

File: Apps/Database/src/Models/UserProfile.cs (hunk 12 above)
Comment (leobcgov) [reply Apps/Database/src/Models/UserProfile.cs]: removed from my last checkin

**Comment by leobcgov** [comment]
finally, I have regenerated the migration scripts to make sure that "removing the virtual list<...> from the existing models" still works.

File: Apps/Database/src/Models/CommunicationEmail.cs [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+        [Key]
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
+        [Column("CommunicationEmailId")]
+        public Guid Id { get; set; }
+
+        /// <summary>
+        /// Gets or sets the Communication Id (foreign key).
+        /// </summary>
+        [ForeignKey("Communication")]
+        public Guid CommunicationId { get; set; }
```
Comment (sslaws) [inline Apps/Database/src/Models/CommunicationEmail.cs:39]: Sorry, this should be required.

File: Apps/Database/src/Models/CommunicationEmail.cs [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+        /// <summary>
+        /// Gets or sets the communication related to this Communication Email.
+        /// </summary>
+        public virtual Communication Communication { get; set; } = null!;
+
+        /// <summary>
+        /// Gets or sets the user profile hdid (foreign key).
+        /// </summary>
+        [MaxLength(52)]
+        [ForeignKey("UserProfile")]
```
Comment (sslaws) [inline Apps/Database/src/Models/CommunicationEmail.cs:50]: and this should be required.

File: Apps/Database/src/Models/CommunicationEmail.cs [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+        /// <summary>
+        /// Gets or sets the user profile related to this Communication Email.
+        /// </summary>
+        public virtual UserProfile UserProfile { get; set; } = null!;
+
+        /// <summary>
+        /// Gets or sets the Email Id (foreign key).
+        /// </summary>
+        [ForeignKey("Email")]
```
Comment (sslaws) [inline Apps/Database/src/Models/CommunicationEmail.cs:61]: and this should be required.

File: Apps/Database/src/Models/CommunicationEmail.cs (hunk 15 above)
Comment (leobcgov) [reply Apps/Database/src/Models/CommunicationEmail.cs:40]: Thanks @sslaws , I have added these attributes and performed rollback/regeneration of db scripts including testing. Can you please review?

File: Apps/Database/src/Migrations/20200724041910_AddCommunicationEmailEntityAndCommunicationProcessingStatus.cs [hunk 18]
```diff
@@ -1,4 +1,21 @@
-﻿using System;
```
Comment (sslaws) [inline Apps/Database/src/Migrations/20200724041910_AddCommunicationEmailEntityAndCommunicationProcessingStatus.cs]: Why does this file have Updates to all of the template data?

**Comment by leobcgov** [comment]
@sslaws per my local tests, the new migration scripts that you have generated working perfectly. Thanks.

**Pull request merged**